### PR TITLE
[axmszr] iP

### PR DESCRIPTION
**Pull request by axmszr**
# Cbot
### Upgrade your to-do list, today.
---

> `System.out.println("Didn't I do it for you?")` ~Siri

Is speaking too much of a hassle? Does your voice-recognition not recognize your voice? Really wish you could go back to the ol' typing days?
Well, here's the solution you need! Meet **Cbot**, your very own text-based chatbot secretary.
```java
  _____  _            _           |
 /  ___]| |   ^-^   _| |_         |        for (Person p : everyone) {
|  / :D | |___  ___[_ + _]        |            cbot.hug(p);    // cbot says hi!
|  \___ | / . \/ . \ | |          |        }
 \_____]|_,_*_/\_*_/ |_/          |
```
With **Cbot**, you can:
1. Keep track of your to-do list, including deadlines and events!
* *Filter*, *search*, or *sort* your tasks, by:
   - date
   - name
   - type
   - completion status
   - and more!
- [ ] Remain organized and consistent. 😄 🧑‍🎓 📔
- [x] Sleep in peace, knowing **Cbot** will take care of you :")
- [ ] ~Ace your university modules~

___

**Cbot** is easy to learn, and easy to use.

Try it now [here](https://www.youtube.com/watch?v=eN6jkWxxm2Y)[^1]!

[^1]: Sorry, try [this](https://www.youtube.com/watch?v=dQw4w9WgXcQ) instead.

### CODE REVIEW

File: src/main/java/cbot/Cbot.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    private final FileStuff fs;
+    private boolean doBye;
+    private boolean prevBad;
+
+    /**
+     * Constructs a fresh Cbot instance.
+     */
+    public Cbot() {
+        this.fs = new FileStuff(PATH);
+        this.doBye = false;
```
Comment (Karen-Lam) [inline src/main/java/cbot/Cbot.java]: Since doBye holds a boolean variable, I think it might be nice to use a variable name that suggests its a boolean (e.g. isDoBye, isBye)

File: src/main/java/cbot/task/TaskList.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+/**
+ * Stores and manages a list of tasks.
+ *
+ * @see Task
+ */
+public class TaskList {
+    //private static final int GAPS = 3;
+    private static final String GAP = "   ";
+
+    private final ArrayList<Task> tdl;
```
Comment (Karen-Lam) [inline src/main/java/cbot/task/TaskList.java]: I think tdl should be in caps, because it is declared as a final variable

File: src/main/java/cbot/task/Task.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+ * @see Deadline
+ * @see Event
+ */
+public class Task implements Comparable<Task> {
+    public static final String DONE_TRUE = "X";
+    public static final String DONE_FALSE = " ";
+    public static final String SEP = " ;; ";
+    public static final String TODO_SYMBOL = "T";
+
+    private final String desc;
```
Comment (Karen-Lam) [inline src/main/java/cbot/task/Task.java]: I think desc should be DESC because it is declared as a final variable

File: src/main/java/cbot/io/UI.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+/**
+ * Manages printing to and receiving from the user, mainly via the Command Line Interface.
+ */
+public class UI {
+
+    //CHECKSTYLE.OFF: SingleSpaceSeparator
+
+    // Frequently Used Strings
+    private static final String BLANK     = "   ";
```
Comment (Karen-Lam) [inline src/main/java/cbot/io/UI.java]: I like how the commonly used Strings are all collated at the top of the code. It makes the code look a lot neater.

File: src/main/java/cbot/io/Command.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+    /**
+     * Returns true if the start of the given String matches the Command keyword.
+     * e.g. "mark 1" matches the keyword "mark " in MARK.
+     *
+     * @param input The text to check.
+     * @return Whether the text means to call this Command.
+     * @see Parser
+     */
+    boolean match(String input) {
```
Comment (Karen-Lam) [inline src/main/java/cbot/io/Command.java]: May be can change the name is isMatch, so it sounds more like a boolean too.

File: src/main/java/cbot/io/Command.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    /**
+     * Returns true if the Command expects body text but receives none.
+     * e.g. "mark  " with no task number specified.
+     *
+     * @param text The body text to check.
+     * @return Whether there is missing body text, given the Command type.
+     */
+    boolean missingText(String text) {
```
Comment (Karen-Lam) [inline src/main/java/cbot/io/Command.java]: May be can change the name is isMissingText, so it sounds more like a boolean too.

**Review by Karen-Lam — COMMENTED** [review]
Looks great! Found your code was easily to read and follow through. I noted some minor violations in coding standards. but other than that, I think the code is great.

**Comment by axmszr** [comment]
Hi @Karen-Lam ! Thank you so much for the comments. I'll definitely look through the booleans and see whether I can name them better!

As for `tdl` and `desc`, I didn't capitalize them because they aren't `static` variables. I think `final` variables don't need to be capitalized by convention? (These names passed the checkstyle. Only `static` variables in camelCase did not.)

File: src/main/java/cbot/io/Parser.java [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+                throw new BadInputException("Invalid index!");
+            }
+            break;
+
+        case UNMARK:
+            try {
+                int num = Integer.parseInt(this.text);
+
+                if (tl.notInRange(num)) {
+                    throw new PoorInputException(tl.rangeError(num));
```
Comment (bokung) [inline src/main/java/cbot/io/Parser.java]: This part seems very heavily indented. Perhaps the overall structure can be reorganised?

File: src/main/java/cbot/io/Parser.java [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+
+        case FILTER:
+            if (tl.getCount() == 0) {
+                throw new PoorInputException("Eh? You have no tasks to filter");
+            }
+
+            String msg;
+            ArrayList<String> arrFilter;
+
+            switch (this.text.toLowerCase()) {
```
Comment (bokung) [inline src/main/java/cbot/io/Parser.java]: Is there a point here in including cases with no behaviour if they are gonna fallthrough anyways?

File: src/main/java/cbot/io/Parser.java [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+                        + UI.printMany(tl.listTasks());
+            }
+            break;
+
+        case MARK:
+            try {
+                int num = Integer.parseInt(this.text);
+
+                if (tl.notInRange(num)) {
+                    throw new PoorInputException(tl.rangeError(num));
```
Comment (bokung) [inline src/main/java/cbot/io/Parser.java]: Same here with the indentation

File: src/main/java/cbot/io/Parser.java [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+import cbot.task.Event;
+import cbot.task.Task;
+import cbot.task.TaskList;
+import cbot.time.TimeStuff;
+
+/**
+ * Handles much of the parsing of user inputs.
+ */
+public class Parser {
+    private Command c;
```
Comment (bokung) [inline src/main/java/cbot/io/Parser.java]: Perhaps this variable can be renamed to be more descriptive? It is used throughout the entire class and it is not ideal to have to refer back constantly to see what `c` is.

File: src/main/java/cbot/io/Parser.java [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+            break;
+
+        case BEFORE:
+            if (tl.getCount() == 0) {
+                throw new PoorInputException("Eh? You have no tasks to filter");
+            }
+
+            LocalDateTime bef = TimeStuff.parseDT(this.text);
+
+            ArrayList<String> arrBef = tl.listFilter(t ->
```
Comment (bokung) [inline src/main/java/cbot/io/Parser.java]: This expression seems pretty complicated, perhaps it can be broken down into more intermediate values?

File: src/main/java/cbot/task/Task.java [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+            return true;
+        }
+    }
+
+    /**
+     * Marks the task as not-yet-done.
+     *
+     * @return Whether the status changed from done to not done.
+     */
+    public boolean noDo() {
```
Comment (bokung) [inline src/main/java/cbot/task/Task.java]: Here i believe this method should be named to clearly reflect what it's doing. Also, this part can be replaced with the following and it will look much neater.
```java
boolean b = this.isDone;
this.isDone = false;
return b;

```

File: src/main/java/cbot/task/Task.java [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        return String.format("%s%s%s%s%s",
+                getSymbol(), SEP, getStatus(), SEP, getDesc());
+    }
+
+    /**
+     * Marks the task as done.
+     *
+     * @return Whether the status changed from not done to done.
+     */
+    public boolean yesDo() {
```
Comment (bokung) [inline src/main/java/cbot/task/Task.java]: Again this method seems to not reflect it's purpose clearly.

File: src/main/java/cbot/io/Parser.java [hunk 14]
```diff
@@ -0,0 +1,362 @@
...
+    /**
+     * Processes the Command and the rest of the input. If needed, the TaskList is modified and the UI is given
+     * outputs to print. The main bulk of Cbot's parsing is done here.
+     *
+     * @param tl The current list of tasks.
+     * @return The String response to the command.
+     * @throws PoorInputException If the input text is improper or erroneous.
+     * @throws DateTimeParseException If some provided datetime is not in a recognized format.
+     */
+    public String respond(TaskList tl) throws PoorInputException, DateTimeParseException {
```
Comment (bokung) [inline src/main/java/cbot/io/Parser.java]: This huge method can perhaps be broken down into smaller functions. It's quite hard to follow the logic of this very long method and you have to indent quite heavily to implement all the logic.

**Review by bokung — COMMENTED** [review]
Overall very well documented code! Good use of OOP in structuring your overall program.

File: src/main/java/cbot/io/Parser.java (hunk 14 above)
Comment (axmszr) [reply src/main/java/cbot/io/Parser.java]: Hm, I considered turning each Command into its own Class, and have the functionality be within those, but that felt like too much work hah
Thank you for pointing this out! I'll definitely think about what I can do to make it more readable.

File: src/main/java/cbot/io/Parser.java (hunk 7 above)
Comment (axmszr) [reply src/main/java/cbot/io/Parser.java]: Hi! Thanks so much for the comments.
Could I ask what you're referring to here? Just this `case UNMARK:` or the whole method? I'm not too sure how I can reorganize this `try`-`catch` block :(

File: src/main/java/cbot/io/Parser.java (hunk 8 above)
Comment (axmszr) [reply src/main/java/cbot/io/Parser.java]: Oh! These are just to give the user a bit more flexibility in how they call the filter. For example, both `filter deadline` and `filter dl` (which do the same thing) are accepted and understood by Cbot.
If you know a neater way of doing this, please do let me know!

**Comment by axmszr** [comment]
Hi @bokung ! Thank you so much for the comments. They were all really helpful, and I have ideas on how to improve some of it.
